x-pack/legacy/plugins/index_management/public/app/components/mappings_editor/lib/mappings_validator.test.ts

```diff
@@ -0,0 +1,309 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { validateMappings, validateProperties, isObject } from './mappings_validator';
+
+jest.mock('ui/index_patterns', () => ({
+  ILLEGAL_CHARACTERS: '',
+  validateIndexPattern: () => ({}),
+}));
+
+describe('Mappings configuration validator', () => {
+  it('should convert non object to empty object', () => {
+    const tests = ['abc', 123, [], null, undefined];
+
+    tests.forEach(testValue => {
+      const { value, errors } = validateMappings(testValue as any);
+      expect(isObject(value)).toBe(true);
+      expect(errors).toBe(undefined);
+    });
+  });
+
+  it('should strip out unknown configuration', () => {
+    const mappings = {
+      dynamic: true,
+      date_detection: true,
+      numeric_detection: true,
+      dynamic_date_formats: ['abc'],
+      _source: {
+        enabled: true,
+        includes: ['abc'],
+        excludes: ['abc'],
+      },
+      properties: { title: { type: 'text' } },
+      unknown: 123,
+    };
+
+    const { value, errors } = validateMappings(mappings);
+
+    const { unknown, ...expected } = mappings;
+    expect(value).toEqual(expected);
+    expect(errors).toBe(undefined);
+  });
+
+  it('should strip out invalid configuration', () => {
+    const mappings = {
+      dynamic: true,
+      numeric_detection: 123, // wrong format
+      dynamic_date_formats: false, // wrong format
+      _source: {
+        enabled: true,
+        includes: 'abc',
+        excludes: ['abc'],
+        wrong: 123, // parameter not allowed
+      },
+      properties: 'abc',
+    };
+
+    const { value, errors } = validateMappings(mappings);
+
+    expect(value).toEqual({
+      dynamic: true,
+      properties: {},
+    });
+
+    expect(errors).not.toBe(undefined);
+    expect(errors!.length).toBe(3);
+    expect(errors!).toEqual([
+      { code: 'ERR_CONFIG', configName: 'numeric_detection' },
+      { code: 'ERR_CONFIG', configName: 'dynamic_date_formats' },
+      { code: 'ERR_CONFIG', configName: '_source' },
+    ]);
+  });
+});
+
+describe('Properties validator', () => {
+  it('should convert non object to empty object', () => {
```

Comment:
I have the same question here about the behavior that we're testing and why it's desirable.

Reply:
Same as above. If this is passed in the JSON

```
{
  "properties": {
    "title": "I am a string"
  }
}
```

Then the validation will deal with it correctly.